### PR TITLE
KDL IK plugin cleanup

### PR DESCRIPTION
**Pull request by rhaschke**
This is an attempt to cleanup and improve the KDL IK plugin.

The first 4 commits simply cleanup code (hopefully) without changing semantics.
326a76fbf6e8c6b664e3c139fd062ebd38b78f7c improves IK when passing through singularities. In that case, the step size of the velocity IK solver needs to be decreased to match the highly nonlinear shape of the FK manifold.

The Kinematics plugin has a major design flaw, which I want to discuss here as well:
Currently it's only possible to specify a timeout for the solver. There is no way to specify the number of search attempts (randomly sampling e.g. a new seeds).
However, [`RobotState::setFromIK`](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_core/robot_state/src/robot_state.cpp#L1422) provides an `attempts` argument. But, even if you specify 1 there, the underlying Kinematics plugin will randomly sample seeds until the `timeout` is reached, which is really annoying if one just wants to test IK for the current seed.

This highlights another design flaw: Both `RobotState::setFromIK` and the Kinematics plugin perform random seed sampling. I think, this should be done only once! I discussed this issue shortly with @v4hn already and we agreed that the random sampling should be removed from `RobotState`, but only be done in the IK plugins (`TracIK` relies on sampling inside the plugin). In that case, the `KinematicsBase` class should provide a utility function to sample new seeds to avoid reimplementing this functionality in all plugin instances.

However, we need a mechanism to pass a `max attempts` argument. I'm afraid that all clean solutions will change API. I see the following options:

- Add an appropriate argument to the function signatures
- Add an appropriate parameter to `KinematicsQueryOptions`
- Some crude hack, passing a special timeout (e.g. negative) to indicate the intend

What do you think?

### CODE REVIEW

**Comment by davetcoleman** [comment]
This is great!

> match the highly nonlinear shape of the FK manifold.

What is a forward kinematics manifold?

> the underlying Kinematics plugin will randomly sample seeds until the timeout is reached, 

KDL does not randomly sample seeds AFAIK, only MoveIt!'s wrapper as you pointed out. I'm confused by your statement here

> Both RobotState::setFromIK and the Kinematics plugin to random seed sampling. I think, this should be done only once!

Which kin plugin are you talking about?

> we agreed that the random sampling should be removed from RobotState

Agreed

> Add an appropriate argument to the function signatures

I like this option the best... the intent of that struct is to allow new options to be added.

I also think Melodic would be a perfect time to do a major cleanup of KinematicsBase and just outline the changes in the migration plan. Want to writeup a draft for the new interface?

File: moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h [hunk 1]
```diff
@@ -190,15 +190,15 @@ class KDLKinematicsPlugin : public kinematics::KinematicsBase
 
   robot_model::RobotModelPtr robot_model_;
 
-  robot_state::RobotStatePtr state_, state_2_;
+  robot_state::RobotStatePtr state_;
```
Comment (davetcoleman) [inline moveit_kinematics/kdl_kinematics_plugin/include/moveit/kdl_kinematics_plugin/kdl_kinematics_plugin.h]: lol

File: moveit_core/robot_state/src/robot_state.cpp [hunk 2]
```diff
@@ -595,54 +595,33 @@ void moveit::core::RobotState::updateLinkTransforms()
 void moveit::core::RobotState::updateLinkTransformsInternal(const JointModel* start)
 {
   const std::vector<const LinkModel*>& links = start->getDescendantLinkModels();
-  if (!links.empty())
+  for (const LinkModel* link : links)
```
Comment (davetcoleman) [inline moveit_core/robot_state/src/robot_state.cpp]: can you add some comments to this function?

File: moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_pos_nr_jl_mimic.cpp [hunk 3]
```diff
@@ -110,79 +110,81 @@ void ChainIkSolverPos_NR_JL_Mimic::qMimicToq(const JntArray& q, JntArray& q_resu
...
-    for (std::size_t i = 0; i < 6; ++i)
-      ROS_DEBUG_NAMED("kdl", "%d: %f", (int)i, delta_twist(i));
-
-    iksolver.CartToJnt(q_temp, delta_twist, delta_q);
-
-    Add(q_temp, delta_q, q_temp);
+    if (delta_twist_err >= last_delta_twist_err) {
+      // if the error increased, we are close to a singularity -> reduce step size
+      double old_step_size = step_size;
+      step_size *= 0.5; // reduce scale;
```
Comment (davetcoleman) [inline moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_pos_nr_jl_mimic.cpp]: seems to me there should be a min step_size that should be checked s.t. the value can't get too close to zero. does that make sense here?

**Comment by davetcoleman** [comment]
Have you checked this PR against the tests in [moveit_kinematic_tests?](https://github.com/ros-planning/moveit_kinematics_tests/issues/12)

**Comment by rhaschke** [comment]
> What is a forward kinematics manifold?

Just the manifold of Cartesian poses spanned by the joint space. Close to a singularity, this manifold has a strong curvature and Jacobian-based IK approaches tend to overshoot.

> KDL does not randomly sample seeds AFAIK.

It does, namely via [`KDLKinematicsPlugin::getRandomConfiguration`](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_kinematics/kdl_kinematics_plugin/src/kdl_kinematics_plugin.cpp#L81) used [here](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_kinematics/kdl_kinematics_plugin/src/kdl_kinematics_plugin.cpp#L515).

> Have you checked this PR against the tests in moveit_kinematic_tests?

Now, I did. All tests pass, except one. I need to look into this.

File: moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_pos_nr_jl_mimic.cpp (hunk 3 above)
Comment (rhaschke) [reply moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_pos_nr_jl_mimic.cpp]: There is a check a few lines down: `if (step_size < eps)`.

**Comment by rhaschke** [comment]
@davetcoleman I already have prepared locally a branch targeted to Melodic for this.
Unfortunately, I didn't find time yet to look into the test regressions. But I suggest to continue this work for Melodic. What are your (short-term) plans for this as you merged kinetic-devel?

**Comment by rhaschke** [comment]
You merged kinetic-devel into this PR branch. I will close this for now and re-open for Melodic.

**Pull request closed without merging**